packages/aws-cdk-lib/aws-ec2/lib/vpc.ts

```diff
@@ -1646,24 +1761,60 @@ export class Vpc extends VpcBase {
         // For reserved azs, do not create any resources
         return;
       }
+      let subnetProps: SubnetProps;
+      if (this.vpcProtocol === VpcProtocol.IPV4_ONLY) {
+        // mapPublicIpOnLaunch true in Subnet.Public, false in Subnet.Private or Subnet.Isolated.
```

Comment:
Extract all this logic out into a helper function. This is an annoying place to stick it all. 

```ts
private calculateMapPublicIpOnLaunch(config: SubnetConfig): boolean | undefined;


const subnetProps = {
  mapPublicIpOnLaunch: this.calculateMapPublicIpOnLaunch(subnetConfig),
  // ...etc...
} satisfies SubnetProps;
```